src/mongocrypt-kms-ctx.c

```diff
@@ -995,6 +1091,30 @@ static bool _ctx_done_kmip_decrypt(mongocrypt_kms_ctx_t *kms_ctx) {
     return ret;
 }
 
+bool mongocrypt_kms_ctx_fail(mongocrypt_kms_ctx_t *kms) {
+    if (!kms || !kms->retry_enabled) {
+        return false;
+    }
+
+    if (kms->attempts >= kms_max_attempts) {
+        mongocrypt_status_t *status = kms->status;
+        CLIENT_ERR("KMS request failed after %d retries due to a network error", kms_max_attempts);
+        return false;
+    }
+
+    /* Not idempotent and can't be retried. */
```

Comment:
Suggest using an allow-list (rather than a block-list) to determine eligibility. This may help to prevent retrying non-idempotent request added in the future.